File: haskell/src/Huffman.hs
{-# LANGUAGE TupleSections #-}

module Huffman
  ( frequencies
  , encode
  , decode
  , Bit(..)
  ) where

import           Data.Map                       ( Map )
import qualified Data.Map                      as M
                                                ( fromListWith
                                                , toList
                                                , fromList
                                                , lookup
                                                )

import           Data.Sequence                  ( Seq(..)
                                                , (|>)
                                                )
import qualified Data.Sequence                 as S
                                                ( fromList
                                                , empty
                                                , sortOn
                                                )

data Bit = Z | O deriving (Eq, Show)

data Tree a = Leaf { weight :: Int, symbol :: a }
            | Node { weight :: Int, left :: Tree a, right :: Tree a }

type Queue a = Seq (Tree a)

-- | Pop head of one of given queues whichever contains lowest-weighted tree.
-- | Returns the minimum tree and updated queue state.
-- | 
-- | Note: By the Huffman tree construction, both queues being empty is an
-- | illegal (and unreachable) state.
popMin :: Queue a -> Queue a -> (Tree a, Queue a, Queue a)
popMin (n :<| nq) Empty      = (n, nq, Empty)
popMin Empty      (l :<| lq) = (l, Empty, lq)
popMin nq@(n :<| nq') lq@(l :<| lq') | weight l <= weight n = (l, nq, lq')
                                     | otherwise            = (n, nq', lq)
popMin _ _ = error "Error: both queues are empty"

-- | Apply single step of the Huffman tree building procedure:
-- |  1. Pop two nodes (or leafs) from the queues with the minimum weights
-- |  2. Construct new node with weights added and the two nodes as sub-trees
-- |  3. Push new node to the back of the node queue
step :: (Queue a, Queue a) -> (Queue a, Queue a)
step (nq, lq) = (nq'' |> n, lq'')
 where
  (l, nq' , lq' ) = popMin nq lq
  (r, nq'', lq'') = popMin nq' lq'
  n               = Node { weight = weight l + weight r, left = l, right = r }

-- | Build the Huffman coding tree given symbol frequencies:
-- |  1. Create a leaf for each symbol and weight in given frequencies
-- |  2. Craate a leaf queue sorted by weight and an initially empty node queue
-- |  3. Repeat `step` until node queue contains single tree
-- |  4. Return final node as the root of the Huffman coding tree
huffmanTree :: [(a, Int)] -> Tree a
huffmanTree = build S.empty . S.sortOn weight . S.fromList . fmap leaf
 where
  leaf (s, w) = Leaf { weight = w, symbol = s }
  build (tree :<| Empty) Empty = tree
  build nq               lq    = let (nq', lq') = step (nq, lq) in build nq' lq'

-- | Calculate symbol frequencies of a text.
frequencies :: Ord a => [a] -> [(a, Int)]
frequencies = M.toList . M.fromListWith (+) . map (, 1)

-- | Encode a sequence using the given frequencies.
encode :: Ord a => [(a, Int)] -> [a] -> Maybe [Bit]
encode []    _       = Nothing
encode [_]   _       = Nothing
encode freqs symbols = concat <$> traverse encodeSymbol symbols
 where
  table = coding $ huffmanTree freqs
  encodeSymbol s = M.lookup s table

-- | Associate each symbol to a coding resulting from given coding tree.
coding :: Ord a => Tree a -> Map a [Bit]
coding t = M.fromList (collect t [] [])
 where
  collect Leaf { symbol = s } bs items = (s, bs) : items
  collect Node { left = l, right = r } bs items =
    collect l (O : bs) (collect r (Z : bs) items)

-- | Decode a bit sequence using the given frequencies.
decode :: Ord a => [(a, Int)] -> [Bit] -> Maybe [a]
decode []    _    = Nothing
decode [_]   _    = Nothing
decode freqs bits = Just (decodeBits bits tree)
 where
  tree = huffmanTree freqs
  decodeBits bs       Leaf { symbol = s } = s : decodeBits bs tree
  decodeBits (O : bs) Node { left = l }   = decodeBits bs l
  decodeBits (Z : bs) Node { right = r }  = decodeBits bs r
  decodeBits []       _                   = []

File: python/codewars/huffman.py
from collections import Counter, deque
from dataclasses import dataclass
from functools import cached_property
from itertools import starmap
from typing import Collection, Deque, Dict, List, Optional, Tuple, Union

Freqs = Collection[Tuple[str, int]]
Tree = Union["Leaf", "Node"]


@dataclass
class Leaf:
    weight: int
    symbol: str
    pred: Optional["Node"]

    @staticmethod
    def new(symbol: str, weight: int) -> "Leaf":
        return Leaf(weight=weight, symbol=symbol, pred=None)

    @cached_property
    def bits(self) -> str:
        bits = []
        node: Tree = self
        while node.pred is not None:
            n = node.pred
            bits.append(int(node is n.right))
            node = n

        return "".join(str(b) for b in reversed(bits))


@dataclass
class Node:
    weight: int
    left: Tree
    right: Tree
    pred: Optional["Node"]

    @staticmethod
    def new(left: Tree, right: Tree) -> "Node":
        node = Node(
            weight=left.weight + right.weight,
            left=left,
            right=right,
            pred=None,
        )
        left.pred = node
        right.pred = node
        return node


@dataclass(frozen=True)
class HuffmanTree:
    root: Node
    leafs: Dict[str, Leaf]

    @classmethod
    def build(cls, freqs: Freqs) -> Optional["HuffmanTree"]:
        freqs = sorted(freqs, key=lambda pair: pair[1])
        return cls.build_presorted(freqs)

    @classmethod
    def build_presorted(cls, freqs: Freqs) -> Optional["HuffmanTree"]:
        if len(freqs) < 2:
            return None

        leaf_refs: Dict[str, Leaf] = {}

        leafs = deque(starmap(Leaf.new, freqs), maxlen=len(freqs))
        nodes: Deque[Node] = deque(maxlen=len(leafs) // 2)

        def min_queue(
            lq: Deque[Leaf], nq: Deque[Node]
        ) -> Union[Deque[Leaf], Deque[Node]]:
            if not lq:
                return nq
            if not nq:
                return lq
            return lq if lq[0].weight <= nq[0].weight else nq

        while len(leafs) + len(nodes) > 1:
            left = min_queue(leafs, nodes).popleft()
            if isinstance(left, Leaf):
                leaf_refs[left.symbol] = left

            right = min_queue(leafs, nodes).popleft()
            if isinstance(right, Leaf):
                leaf_refs[right.symbol] = right

            nodes.append(Node.new(left, right))

        assert not leafs and len(nodes) == 1
        return cls(root=nodes[0], leafs=leaf_refs)

    def encode(self, symbol: str) -> str:
        return self.leafs[symbol].bits


def frequencies(s: str) -> Freqs:
    return Counter(s).items()


def encode(freqs: Freqs, s: str) -> Optional[str]:
    tree = HuffmanTree.build(freqs)

    if tree is None:
        return None

    return "".join(tree.encode(c) for c in s)


def decode(freqs: Freqs, bits: str) -> Optional[str]:
    tree = HuffmanTree.build(freqs)

    if tree is None:
        return None

    chars: List[str] = []
    node = tree.root

    for bit in bits:
        n = node.right if int(bit) else node.left

        if isinstance(n, Leaf):
            chars.append(n.symbol)
            node = tree.root
        else:
            node = n

    assert node is tree.root
    return "".join(chars)
